### DISCUSSION TITLE
Understanding createEffect with store

### DISCUSSION THREAD

**Discussion by yonathan06** [solidjs/solid#829] [Q&A] [answered]
Following this example: https://stackblitz.com/edit/solid-vite-hhhbce?file=src%2Findex.tsx

I'm mutating the store using `produce`, for example:
```JavaScript
 const onTodoChange = (id: number, value: string) => {
    setStore(
      'todos',
      (todo) => todo.id === id,
      'text',
      () => value
    );
  };

  const addNewTodo = () => {
    setStore(
      'todos',
      produce((todos) =>
        todos.push({ id: todos[todos.length - 1].id + 1, text: '' })
      )
    );
  };
```

And I'm subscribing to store change using `createEffect`:
```JavaScript
createEffect(() => {
    // Works on adding new todo, not when changing todo
    console.log('createEffect store.todos', store.todos);
  });
  createEffect(() => {
    // Doesnt work
    console.log('createEffect store', store);
  });
```

when accessing `store.todo` the effect is triggered when adding a new todo, not when changing one.
the second is not triggered at all.

I guess this is because of fined reactivity (`todos` is a separate reactive proxy than `todos[index].text`), so need to be explicit in the `createEffect`. If so, maybe a utility like `createUnionEffect`/`createDeepEffect` would be helpful?

**Comment by lxsmnsyc** [comment, marked as answer]
The first one doesn't work because you're tracking `store.todos` and not an specific `todo` item inside `store.todos`.

The second one doesn't work because it's a variable, and variables cannot be reactive (hence why we use functions for signals rather than just plain variables).

If the intent is to track the whole store, there are some nifty and cheap workarounds:
- Adding a special attribute to the store that acts like a "version". You then change this value whenever you attempt to change a property of that store.
- Using a signal instead of a store.
- Creating a `deepTrack` utility

```js
function deepTrack(store) {
  for (const k in store) {
    const value = store[k];
    if (typeof value === 'object') {
      deepTrack(store);
    }
  }
}
```

**Reply by yonathan06** [reply]
Thanks! maybe it is worth having `deepTrack` or `createDeepEffect` in solid? or maybe it will be abused 😅

**Reply by lxsmnsyc** [reply]
Not sure how would you implement a `createDeepEffect`.

**Reply by MrDesjardins** [reply]
@LXSMNSYC I wanted to tell you that I stumbled into the same desire as @yonathan06.

My use case is that I started to have too many `createSignal` in a context. I wanted to refactor using a `createStore`. However, there are parts of the system that need to _react_ to a change of the context. It was working well with `createEffect` using the context and the signal. However, when refactoring using the `createStore` the usage of reactivity is lost.

There are definitely scenarios where you want to _react_ to a change in the store and it is surprising to see a departure from the usage of `Accessor` in the signal part of SignaJS compared to the store.

Do you have any alternative solution to be able to "watch" specific properties of a store or is the expectation to rely on signal solely?

**Reply by lxsmnsyc** [reply]
I'm not sure how to answer that. Probably there's a misconception on how stores work and how it should be used. Do you have any minimal repro?

**Reply by MrDesjardins** [reply]
@LXSMNSYC  I do not have a quick repro on hand. I can maybe make one later this week.

In short, something like the following (working with signal). The effect listen to the `userData` which is a context and is listening (`createEffect`) on a signal `selectedCoordinate`. When the selection change, there is a timer being stopped and a reset of colors...

```
  createEffect(
    // When a new pixel is selected, we reset all the colors
    on(userData?.selectedCoordinate ?? [], () => {
      if (userData?.selectedCoordinate !== undefined) {
        clearTimeout(timeOutRef);
        setColors(COLORS);
      }
    }),
  );
```
Now, if we refactor the `userData` to instead of exposing in the context the `Accessor<>` but to expose a property of a state, it does not work because `on` does not work on a store's property... it expects a `Accessor<>`.

In React we can easily refactor code from useState to useReducer; in that case using a `useEffect` with a reference (second parameter) to the property (in both cases state or reducer).

**Reply by lxsmnsyc** [reply]
`Accessor` is basically just `() => T` so you can solve your problem by turning `on(userData?.selectedCoordinate ?? [], ...)` into `on(() => userData?.selectedCoordinate ?? [], ...)`.

The thing is you are reading `userData?.selectedCoordinate` static and by static I mean only exactly once. To help understand this, we can separate your code into the ff:

```js
// Do take note that components in Solid only evaluate ONCE unlike in React
const tracked = userData?.selectedCoordinate ?? []; // This is read statically since you are accessing the value only during the component's evaluation
const handler = on(tracked, () => {
  if (userData?.selectedCoordinate !== undefined) {
    clearTimeout(timeOutRef);
    setColors(COLORS);
  }
});
createEffect(handler);
```

`on` accepts an `Accessor` for this purpose, as reactivity can come from anywhere not just signals.

**Reply by MrDesjardins** [reply]
Thank you @LXSMNSYC Your response not only fixed my issue but also enlighten unclear conceptions around the Accessor wrapper.

A slight modification from your snippet was required but was easy to figure out with your clear explanation

```
const tracked = () => userData?.selectedCoordinate ?? [];
```

Thank you, have a great evening

**Reply by yw662** [reply]
Shouldn't it be useful is the first version can work? There can be, maybe some "bubbling" when an item inside a store is changed.

**Comment by yonathan06** [comment]
Created a small utility that wraps the `setStore` function, and updates a signal as a side effect that can be subscribed to later:

```JavaScript
const [store, setStoreOrigin] = createStore<LearningModule>(initialModules);
const [triggerUpdate, setTriggerUpdating] = createSignal(1);
const setStore: typeof setStoreOrigin = (...args) => {
    setTriggerUpdating((v) => v + 1);
    setStoreOrigin(...args);
};

 createEffect(() => {
    console.log(triggerUpdate());
    console.log(unwrap(store));
});
```

**Reply by azmy60** [reply]
I'm getting an error that says `rest parameter args implicitly has an 'any[]' type.`. I know I can easily fix this by just adding `// @ts-ignore`, but just out of curiosity, does anybody know how to solve this without ignoring the error?

**Reply by yonathan06** [reply]
You can obtain the parameter type:
```TS
const [store, setStoreOrigin] = createStore<LearningModule>(initialModules);
const [triggerUpdate, setTriggerUpdating] = createSignal(1);
const setStore: typeof setStoreOrigin = (...args: Parameters<typeof setStoreOrigin>) => {
    setTriggerUpdating((v) => v + 1);
    setStoreOrigin(...args);
};
```

**Reply by azmy60** [reply]
Thanks for the reply! The error now shows here 🤔 
```ts
const setStore: typeof setStoreOrigin = (...args: Parameters<typeof setStoreOrigin>) => {
    setTriggerUpdating((v) => v + 1);
    setStoreOrigin(...args); // A spread argument must either have a tuple type or be passed to a rest parameter
   //                ^^^^^
};
```

**Reply by yw662** [reply]
looks like for me an issue of typescript

**Reply by yonathan06** [reply]
try removing the `typeof setStoreOrigin`
